feat(rbac): display administration to authorized users

debsmita1 opened the pull request:
**Resolves:**
https://github.com/janus-idp/backstage-plugins/issues/842

**Test setup:**
```app-config.yaml```
```
auth:
  providers:
    github:
      development:
        clientId: $GITHUB_CLIENT_ID
        clientSecret: $GITHUB_CLIENT_SECRET
        
permission:
  enabled: true
  rbac:
    admin:
      users:
        - name: user:default/<your-github-id>
        - name: group:default/admins
```

```DynamicRoot.tsx```
```
Add SignInPage component 

    if (!app.current) {
      app.current = createApp({
        apis,
        components: {
          SignInPage: props => (
            <SignInPage
              {...props}
              auto
              provider={{
                id: 'github-auth-provider',
                title: 'GitHub',
                message: 'Sign in using GitHub',
                apiRef: githubAuthApiRef,
              }}
            />
          ),
        },
```

```auth.ts```
```
      github: providers.github.create({
        signIn: {
          resolver({ result }, ctx) {
            const userRef = `user:default/${result.fullProfile.username}`;
            return ctx.issueToken({
              claims: {
                sub: userRef, // The user's own identity
                ent: [userRef], // A list of identities that the user claims ownership through
              },
            });
          },
          // resolver: providers.github.resolvers.usernameMatchingUserEntityName(),
        },
      }),
```


**_To load the Administration sidebar item dynamically_**

Test Setup

Add the `@janus-idp/backstage-plugin-rbac` package in the packages/app/package.json
Keep the app-config changes from the above set-up

app-config.yaml

```
dynamicPlugins:
  frontend:
    janus.dynamic-frontend-plugin:
      dynamicRoutes:
        …
        - path: '/rbac'
          importName: 'RbacPage'
          module: 'Administration'


```

Root.tsx

```
	export const Root = ({ children }: PropsWithChildren<{}>) => (
	…
        <ScalprumComponent
        scope="janus.dynamic-frontend-plugin"
        module="Administration"
        importName="Administration"
      />
	…
```

dynamic-modules-registry.json

```
      {
   	"module": "Administration",
   	 "scope": "janus.dynamic-frontend-plugin",
    	"manifestLocation": "http://localhost:8004/plugin-manifest.json"
  	}
```

dynamic-frontend-plugin/package.json

```
"scalprum": {
    …
    "exposedModules": {
       …
      "Administration": "./src/remotes/Administration"
    }
  }
```

Create Administration/index.ts under dynamic-frontend-plugin/src/remotes

`export * from '@janus-idp/backstage-plugin-rbac';`


**GIF:**

https://github.com/janus-idp/backstage-plugins/assets/22490998/69b8c23a-2f74-4db0-8839-222748f40f66

Updated the Administration Nav icon
![Screenshot 2023-11-07 at 10 19 18 PM](https://github.com/janus-idp/backstage-plugins/assets/22490998/ecefd61a-9f81-4f74-a5f8-9f2f8a180894)

invincibleJai commented on packages/app/src/App.tsx:
```diff
@@ -66,6 +69,17 @@ export const AppBase = () => {
               {searchPage}
             </Route>
             <Route path="/catalog-graph" element={<CatalogGraphPage />} />
+            <Route
+              path="/rbac"
+              element={
+                <RequirePermission
+                  permission={policyEntityReadPermission}
+                  resourceRef="policy-entity"
+                >
+                  <RbacPage />
+                </RequirePermission>
+              }
+            />
```
good to add these in README and we can remove these changes in `**/app/**`

invincibleJai replied:
@tumido does the above comment make sense or this is an RBAC plugin so the change is fine?

divyanshiGupta replied:
@invincibleJai your previous comment makes sense as this change will not be released with RBAC plugin as its in the core App component and the user will anyways need to add this rbac <Route /> manually in their App component.

invincibleJai commented:
@debsmita1 and frontend dynamic plugin support is there now so pls ensure we test that flow as well and capture it as part of the PR description

ShiranHi commented:
@debsmita1 looks good to me, thank you!

divyanshiGupta commented on packages/app/src/components/Root/Root.tsx:
```diff
@@ -77,6 +79,7 @@ export const Root = ({ children }: PropsWithChildren<{}>) => (
       </SidebarGroup>
       <SidebarSpace />
       <SidebarDivider />
+      <Administration />
```
https://github.com/janus-idp/backstage-plugins/pull/895/files#r1386635365 - Same applies for this

divyanshiGupta commented on packages/app/package.json:
```diff
@@ -47,6 +47,8 @@
     "@backstage/plugin-techdocs-module-addons-contrib": "^1.1.1",
     "@backstage/plugin-techdocs-react": "^1.1.12",
     "@backstage/plugin-user-settings": "^0.7.11",
+    "@janus-idp/backstage-plugin-rbac": "1.1.2",
+    "@janus-idp/backstage-plugin-rbac-common": "1.0.1",
```
@debsmita1 We havent added other plugins here so do we need to add these?

debsmita1 commented:
@divyanshiGupta @invincibleJai I have removed the changes from the app folder. Added ReadMe. And documented the steps to load the Administration sidebar item dynamically in the ReadMe. PTAL

debsmita1 commented:
Some existing tests are failing, working on them

invincibleJai reviewed (approved):
LGTM

Thanks @debsmita1 , let's fix the test

AndrienkoAleksandr commented:
Build failed. Corresponding fix is here https://github.com/janus-idp/backstage-plugins/pull/929

invincibleJai reviewed (approved):
LGTM
